synapse/handlers/directory.py

```diff
@@ -323,18 +303,36 @@ def _update_canonical_alias(self, requester, user_id, room_id, room_alias):
             room_id, EventTypes.CanonicalAlias, ""
         )
 
+        # There is no canonical alias, nothing to do.
+        if not alias_event:
+            return
+
+        # An updated canonical alias event must be sent if the removed alias was
+        # set as the alias on the canonical_alias event or listed in the
+        # alt_aliases field.
         alias_str = room_alias.to_string()
-        if not alias_event or alias_event.content.get("alias", "") != alias_str:
+        alt_aliases = alias_event.content.get("alt_aliases", [])
+        send_update = (
+            alias_event.content.get("alias", "") == alias_str
+            or alias_str in alt_aliases
+        )
+        if not send_update:
             return
 
+        # Include any remaining aliases in the content.
+        content = {}
+        alt_aliases = [alias for alias in alt_aliases if alias != alias_str]
+        if alt_aliases:
+            content["alt_aliases"] = alt_aliases
```

Comment:
this looks like it will drop `content["alias"]` even if that alias still exists?

Reply:
Ah, in the case that the removed alias is not the primary alias. 👍 Good catch.